pty/src/win/pseudocon.rs

```diff
@@ -65,13 +65,31 @@ lazy_static! {
 
 pub struct PsuedoCon {
     con: HPCON,
+    output: FileDescriptor,
 }
 
 unsafe impl Send for PsuedoCon {}
 unsafe impl Sync for PsuedoCon {}
 
 impl Drop for PsuedoCon {
     fn drop(&mut self) {
+        // NOTE:
+        //  Reading output pipe until it encounters an error or number of bytes read is zero. We
+        //  don't check ERROR_BROKEN_PIPE error, because for regular pipes that's the same anyway,
+        //  but for special pipes that support "graceful shutdown" (like TCP's FIN/ACK), the read
+        //  will be successful while number of bytes read is zero. This would mean that the other
+        //  side wants to end things.
+        //  See https://github.com/microsoft/terminal/discussions/17716#discussioncomment-10341857
+        //  for more context.
+        let mut buffer = [0; 1000];
+
+        // read up to 1000 bytes
+        while let Ok(num_byes_read) = self.output.read(&mut buffer[..]) {
+            if num_byes_read == 0 {
+                break;
+            }
+        }
```

Comment:
This won't work, because until recently you must read from the output pipe concurrently while `ClosePseudoConsole` is called.

However, starting https://github.com/microsoft/terminal/pull/17704 you can do something else: You can first call the new `ClosePseudoConsole` and then (on the same thread; not concurrently) read from the output pipe until it is closed on you (returns an error).

If you don't want to upgrade ConPTY now, you can do something like this as a temporary workaround (pseudo-code; I keep forgetting what valid Rust syntax looks like lol):
```rust
fn drop(&mut self) {
    // A `HPCON` is a struct consisting of 3 HANDLEs:
    // A pipe for communicating with the PTY, a handle to keep it alive
    // until ClosePseudoConsole is called, and a process handle to the
    // underlying conhost process.
    unsafe {
        let handles = self.con as *mut [HANDLE; 3];
        for i in ..3 {
            CloseHandle(handles[i]);
            handles[i] = std::ptr::null();
        }
    }

    // read up to 1000 bytes
    while let Ok(num_byes_read) = self.output.read(&mut buffer[..]) {
        if num_byes_read == 0 {
            break;
        }
    }

    // This won't do anything but deallocate the handle.
    unsafe { (CONPTY.ClosePseudoConsole)(self.con) };
}
```

`ReleasePseudoConsole` is a new function, which does the same as the unsafe code above but only for the 2nd handle (which keeps conhost alive). I've started writing documentation for it here: https://github.com/MicrosoftDocs/Console-Docs/pull/323
Here's a preview of the documentation: https://github.com/MicrosoftDocs/Console-Docs/blob/dev/lhecker/24h2/docs/releasepseudoconsole.md

Wezterm should call `ReleasePseudoConsole` (or close the 2nd HPCON handle) right after attaching an application to ConPTY, about here: https://github.com/FrancescElies/wezterm/blob/a25e7be9bc6c68290ed465b6c1aa780cfee9b793/pty/src/win/pseudocon.rs#L189-L191

Ideally it should also stop returning a process handle from that function and use it to track the shell/application lifetime. Instead, it should read from the output pipe until the pipe has been closed on it by the PTY. If wezterm wants to kill all attached console applications prematurely (e.g. when closing the tab), it can call the upcoming non-blocking `ClosePseudoConsole`, or it can close all 3 handles like in the example above.

Long term, wezterm should use the proper ConPTY APIs of course, but short term it would be absolutely fine to cost `HPCON` to an array of handles. It's struct layout is part of the kernel ABI and so it won't change in any existing version and is unlikely to change any time soon.

Reply:
will read your message in detail and change this pr accordingly on monday when I am back in a windows box

Reply:
> This won't work, because until recently you must read from the output pipe concurrently while `ClosePseudoConsole` is called. This is also true for the version of ConPTY that wezterm uses.

Understood, I was suspecting that, in the docs I read that is recommended to drain
communication channels in a separate thread, but I thought maybe I could get
away without creating a separate thread and keep things simpler.

I couldn't put that block after `ClosePseudoConsole` because that's the
function that was blocking but you solved that problem for me below.

> However, starting [microsoft/terminal#17704](https://github.com/microsoft/terminal/pull/17704) you can do something else: You can first call the new `ClosePseudoConsole` and then (on the same thread; not concurrently) read from the output pipe until it is closed on you (returns an error). Basically, you can do what you wrote, but with the order flipped.

One part of me is eager to upgrade and learn about conpty and wezterm but given
that I don't know wezterm's codebase and only know about conpty what you taught
me I will leave that for a separate pr and keep this pr with the minimal patch
that solves the closing panes hanging problem.

> If you don't want to upgrade ConPTY now, you can do something like this as a temporary workaround (pseudo-code; I keep forgetting what valid Rust syntax looks like lol):
>
> ```rust
> fn drop(&mut self) {
>     // A `HPCON` is a struct consisting of 3 HANDLEs:
>     // A pipe for communicating with the PTY, a handle to keep it alive
>     // until ClosePseudoConsole is called, and a process handle to the
>     // underlying conhost process.
>     unsafe {
>         let handles = self.con as *mut [HANDLE; 3];
>         for i in ..3 {
>             CloseHandle(handles[i]);
>             handles[i] = std::ptr::null();
>         }
>     }
>
>     // read up to 1000 bytes
>     while let Ok(num_byes_read) = self.output.read(&mut buffer[..]) {
>         if num_byes_read == 0 {
>             break;
>         }
>     }
>
>     // This won't do anything but deallocate the handle.
>     unsafe { (CONPTY.ClosePseudoConsole)(self.con) };
> }
> ```

Thanks for the example, for this PR this will be the easiest I will just copy pasta your code and make it compile.

> That aside, `ReleasePseudoConsole` is a new function which does the same as the unsafe code above but only for the 2nd handle (which keeps conhost alive). I've started writing documentation for it here: [MicrosoftDocs/Console-Docs#323](https://github.com/MicrosoftDocs/Console-Docs/pull/323) Here's a preview of the documentation: https://github.com/MicrosoftDocs/Console-Docs/blob/dev/lhecker/24h2/docs/releasepseudoconsole.md
>
> Ideally Wezterm should call `ReleasePseudoConsole` (or close the 2nd HPCON handle) right after attaching an application to ConPTY, about here: https://github.com/FrancescElies/wezterm/blob/a25e7be9bc6c68290ed465b6c1aa780cfee9b793/pty/src/win/pseudocon.rs#L189-L191
>
> It should then stop returning a process handle from that function and not use it to track the shell/application lifetime anymore. Instead, it should read from the output pipe until the pipe has been closed on it by the PTY. If wezterm wants to kill all attached console applications prematurely (e.g. when closing the tab), it can call the upcoming non-blocking `ClosePseudoConsole`, or it can close all 3 handles like in the example above.
>
> Long term, wezterm should use the proper ConPTY APIs of course, but short term it would be absolutely fine to cast `HPCON` to an array of handles. Its struct layout is part of the kernel ABI and so it won't change in any existing version and is unlikely to change any time soon.

Even though your explanations are very clear I will leave that for a separate
PR, I am sure if I attempt this I will misinterpret things and will need you to
hold my hand in the process.

I will try to maximize the chances of just fixing my issue first and later on
attempt that or at least create an issue linking your comment in case someone
else feels like trying the upgrade.